packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -507,17 +534,59 @@ function voidFunctionParams(
       : subType.getConstructSignatures();
     for (const signature of signatures) {
       for (const [index, parameter] of signature.parameters.entries()) {
-        const type = checker.getTypeOfSymbolAtLocation(
+        const decl = parameter.valueDeclaration;
+        let type = checker.getTypeOfSymbolAtLocation(
           parameter,
           node.expression,
         );
-        if (isThenableReturningFunctionType(checker, node.expression, type)) {
-          thenableReturnIndices.add(index);
-        } else if (
-          !thenableReturnIndices.has(index) &&
-          isVoidReturningFunctionType(checker, node.expression, type)
-        ) {
-          voidReturnIndices.add(index);
+
+        // If this is a array 'rest' parameter, check all of the argument indices
+        // from the current argument to the end.
+        // Note - we currently do not support 'spread' arguments - adding support for them
+        // is tracked in https://github.com/typescript-eslint/typescript-eslint/issues/5744
+        if (decl && ts.isParameter(decl) && decl.dotDotDotToken) {
+          if (checker.isArrayType(type)) {
+            // Unwrap 'Array<MaybeVoidFunction>' to 'MaybeVoidFunction',
+            // so that we'll handle it in the same way as a non-rest
+            // 'param: MaybeVoidFunction'
+            type = checker.getTypeArguments(type)[0];
+            for (let i = index; i < node.arguments.length; i++) {
+              checkThenableOrVoid(
+                checker,
+                node,
+                type,
+                i,
+                thenableReturnIndices,
+                voidReturnIndices,
+              );
+            }
+          } else if (checker.isTupleType(type)) {
+            // Check each type in the tuple - for example, [boolean, () => void] would
+            // add the index of the second tuple parameter to 'voidReturnIndices'
+            const typeArgs = checker.getTypeArguments(type);
+            for (let i = index; i < node.arguments.length; i++) {
+              checkThenableOrVoid(
+                checker,
+                node,
+                typeArgs[i - index],
+                i,
+                thenableReturnIndices,
+                voidReturnIndices,
+              );
+            }
+          } else {
+            // We must have an 'any' parameter (e.g. `...args: any`), as anything
+            // else would be invalid TypeScript. There's nothing for us to check.
```

Comment:
Nit: I think this is implied by the code around it. If it's neither an array type nor a tuple type, there's nothing for us to do.

```suggestion
```

Reply:
I've removed this branch